File: devohda/python/PROGRAMMERS 43163.py
from collections import deque


def solution(begin, target, words):
    visited = [False] * len(words)
    queue = deque([(begin, 0)])

    while queue:
        word, cnt = queue.popleft()

        if word == target:
            return cnt

        for i in range(len(words)):
            difference = 0

            if visited[i]:
                continue

            for a, b in zip(word, words[i]):
                if a != b:
                    difference += 1
            if difference == 1:
                visited[i] = True
                queue.append((words[i], cnt + 1))

    return 0

File: devohda/javascript/PROGRAMMERS 43163.js
function solution(begin, target, words) {
    const visited = Array.from({length : words.length}, () => false);
    const queue = [[0, begin]];

    while(queue.length !== 0){
        const [cnt, word] = queue.shift();

        if(word === target){
            return cnt
        }

        for(let i = 0; i < words.length; i++){
            let difference = 0;
            if(visited[i]){
                continue;
            }

            for(let j = 0; j < word.length; j++){
                if(word[j] !== words[i][j]){
                    difference++;
                }
            }

            if(difference === 1){
                visited[i] = true;
                queue.push([cnt + 1, words[i]]);
            }
        }
    }

    return 0;
}